hibernate-jcache/src/main/java/org/hibernate/cache/jcache/access/AbstractReadWriteRegionAccessStrategy.java

```diff
@@ -0,0 +1,404 @@
+	protected final Comparator versionComparator;
+	private final UUID uuid = UUID.randomUUID();
+	private final AtomicLong nextLockId = new AtomicLong();
+	private final AtomicLong nextItemId = new AtomicLong();
+
+	public AbstractReadWriteRegionAccessStrategy(R region) {
+		this.versionComparator = region.getCacheDataDescription().getVersionComparator();
+		this.region = region;
+	}
+
+	public R getRegion() {
+		return region;
+	}
+
+	public Object get(SharedSessionContractImplementor session, Object key, long txTimestamp) throws CacheException {
+		final Lockable item = (Lockable) region.get( key );
+
+		final boolean readable = item != null && item.isReadable( txTimestamp );
+		if ( readable ) {
+			return item.getValue();
+		}
+		else {
+			return null;
+		}
+	}
+
+	public boolean putFromLoad(SharedSessionContractImplementor session, Object key, Object value, long txTimestamp, Object version) throws CacheException {
+		while (true) {
+			Lockable item = (Lockable) region.get( key );
+
+			if (item == null) {
+				/*
+				 * If the item is null due a softlock being evicted... then this
+				 * is wrong, the in-doubt soft-lock could get replaced with the
+				 * old value.  All that can be done from a JCache perspective is
+				 * to log a warning.
+				 */
+				if (region.putIfAbsent( key, new Item( value, version, txTimestamp, nextItemId() ))) {
+					return true;
+				}
+			}
+			else if (item.isWriteable( txTimestamp, version, versionComparator )) {
+				if (region.replace( key, item, new Item( value, version, txTimestamp, nextItemId() ))) {
+					return true;
+				}
+			}
+			else {
+				return false;
+			}
+		}
+	}
+
+	public boolean putFromLoad(SharedSessionContractImplementor session, Object key, Object value, long txTimestamp, Object version, boolean minimalPutOverride)
+			throws CacheException {
+		return putFromLoad( session, key, value, txTimestamp, version );
+	}
+
+	public SoftLock lockItem(SharedSessionContractImplementor session, Object key, Object version) throws CacheException {
+		while (true) {
+			Lockable item = (Lockable) region.get( key );
+			long timeout = region.nextTimestamp() + region.getTimeout();
+
+			if ( item == null ) {
+				/*
+				 * What happens here if a previous soft-lock was evicted to make
+				 * this null.
+				 */
+				Lock lock = new Lock(timeout, uuid, nextLockId(), version);
+				if (region.putIfAbsent( key, lock )) {
+					return lock;
+				}
+			}
+			else {
+				Lock lock = item.lock( timeout, uuid, nextLockId() );
+				if (region.replace(key, item, lock)) {
+					return lock;
+				}
+			}
+		}
+	}
+
+	public void unlockItem(SharedSessionContractImplementor session, Object key, SoftLock lock) throws CacheException {
+		while (true) {
+			Lockable item = (Lockable) region.get( key );
+
+			if (item != null && item.isUnlockable( lock )) {
+				if (region.replace(key, item, ((Lock) item ).unlock(region.nextTimestamp()))) {
+					return;
+				}
+			}
+			else {
+				handleMissingLock( key, item );
```

Comment:
Given the consistency guarantee you want with `READ_WRITE` you can of need this to be there ... for a while ...
So I guess only eviction would end up removing it aside from having it replaced.